lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,298 @@
+module OffsitePayments #:nodoc:
+  module Integrations #:nodoc:
+    module Paytm
+      require 'openssl'
+      require 'base64'
+      require 'digest'
+      require 'securerandom'
+
+      $DEBUG = false
+      #$RUN_UNIT_TEST = 0 #do not make this true, in normal
+
+      mattr_accessor :test_url
+      mattr_accessor :production_url
+
+      self.test_url = 'https://pguat.paytm.com/oltp-web/processTransaction'
+      self.production_url = 'https://secure.paytm.in/oltp-web/processTransaction'
+
+      def self.service_url
+        # self.production_url
+        OffsitePayments.mode == :production ? production_url : test_url
+      end
+
+      def self.notification(post, options = {})
+        Notification.new(post, options)
+      end
+
+      def self.return(post, options = {})
+        Return.new(post, options)
+      end
+
+      def self.checksum(chkParams, secret_key, salt = nil)
+        if salt.nil?
+          o = [('a'..'z'), ('A'..'Z'), ('0'..'9')].map { |i| i.to_a }.flatten
+          salt = (0...4).map { o[Random.rand(o.length)] }.join
+        end
+        keys = chkParams.keys
+        str = nil
+        keystr = nil
+        keys = keys.sort
+
+        keys.each do |k|
+          if chkParams[k].nil?
+            next
+          end
+
+          if str.nil?
+            str = chkParams[k].to_s
+            next
+          end
+          str = str + '|' + chkParams[k].to_s
+        end
+        str = str + '|' + salt
+
+        check_sum = Digest::SHA256.hexdigest(str)
+        check_sum += salt
+
+        ### encrypting checksum ###
+        aes = OpenSSL::Cipher::Cipher.new('aes-128-cbc')
+        aes.encrypt
+        aes.key = secret_key
+        aes.iv = '@@@@&&&&####$$$$'
+
+        encrypted_data = aes.update(check_sum.to_s) + aes.final
+        encrypted_data = Base64.encode64(encrypted_data)
```

Comment:
but it will affect other special characters which we don't want to modify

for example 

with **Base64.urlsafe_encode64**
BTg-oopogadkkHRjUwbHy8lVXxdEjKzwjtPJMc9AdD20PP9xTiGfw2n8ilHDCQemRIGd8ip0CHrsBsBr9Wmv69uB0B4S7c_TW7emkhffhUo=

with **Base64.encode64**
BTg+oopogadkkHRjUwbHy8lVXxdEjKzwjtPJMc9AdD20PP9xTiGfw2n8ilHDCQemRIGd8ip0CHrsBsBr9Wmv69uB0B4S7c/TW7emkhffhUo=